server/server.go

```diff
@@ -1147,3 +1156,188 @@ func (s *Server) getClientConnectURLs() []string {
 
 	return urls
 }
+
+// if the ip is not specified, attempt to resolve it
+func resolveHostPorts(addr net.Listener) []string {
+	hostPorts := make([]string, 0)
+	hp := addr.Addr().(*net.TCPAddr)
+	port := strconv.Itoa(hp.Port)
+	if hp.IP.IsUnspecified() {
+		var ip net.IP
+		ifaces, _ := net.Interfaces()
+		for _, i := range ifaces {
+			addrs, _ := i.Addrs()
+			for _, addr := range addrs {
+				switch v := addr.(type) {
+				case *net.IPNet:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				case *net.IPAddr:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				default:
+					continue
+				}
+			}
+		}
+	} else {
+		hostPorts = append(hostPorts, net.JoinHostPort(hp.IP.String(), port))
+	}
+	return hostPorts
+}
+
+// format the address of a net.Listener with a protocol
+func formatURL(protocol string, addr net.Listener) []string {
+	hostports := resolveHostPorts(addr)
+	for i, hp := range hostports {
+		hostports[i] = fmt.Sprintf("%s://%s", protocol, hp)
+	}
+	return hostports
+}
+
+// Ports describes URLs that the server can be contacted in
+type Ports struct {
+	Nats       []string `json:"nats,omitempty"`
+	Monitoring []string `json:"monitoring,omitempty"`
+	Cluster    []string `json:"cluster,omitempty"`
+	Profile    []string `json:"profile,omitempty"`
+}
+
+// Returns a Ports struct describing ports where the server can be contacted
+func (s *Server) PortsInfo(opts *Options) Ports {
```

Comment:
Again, I would not ask the caller to pass that information. This should be retrieved internally from s.getOpts()